### PR TITLE
OPENEUROPA-2416: Allow to define which view modes are embedable.

### PR DESCRIPTION
**Pull request by imanoleguskiza**
### Description

Add functionality to define whether a media view display can be embeded or not.

### Change log

- Added: Add functionality to define whether a media view display can be embeded or not.
- Changed:
- Deprecated:
- Removed:
- Fixed:
- Security:

### Commands

```sh
[Insert commands here]

```

### CODE REVIEW

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 1]
```diff
@@ -336,18 +348,45 @@ public function buildEmbedStep(array $form, FormStateInterface $form_state): arr
...
+          '#type' => 'link',
+          '#title' => t('Manage @media view modes', ['@media' => $entity->bundle()]),
+          '#url' => Url::fromRoute('entity.entity_view_display.' . $entity->getEntityTypeId() . '.default', [
+            'media_type' => $entity->bundle(),
+          ]),
+          '#attributes' => ['target' => '_blank'],
+        ];;
+      }
+    }
+    else {
```
Comment (sergepavle) [inline modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: Could we use here like 
```suggestion
    elseif(count($display_options) === 1) {
```
for simplicity?

File: modules/oe_media_embed/oe_media_embed.post_update.php [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+declare(strict_types = 1);
+
+use Drupal\Core\Config\FileStorage;
+use Drupal\Core\Entity\Entity\EntityViewDisplay;
+use Drupal\Core\Entity\Entity\EntityViewMode;
+
+/**
+ * Enable default embed view modes.
+ */
+function oe_media_embed_post_update_enable_embed_view_modes(): void {
```
Comment (sergepavle) [inline modules/oe_media_embed/oe_media_embed.post_update.php]: Looks like number of post_update is missed.

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 3]
```diff
@@ -336,12 +348,39 @@ public function buildEmbedStep(array $form, FormStateInterface $form_state): arr
...
     // Allow to specify a view mode if the media type has more than 1.
-    $options = $this->getMediaViewModeOptions($entity);
-    if ($options) {
+    $display_options = $this->getMediaViewModeOptions($entity);
+
+    // If no view mode was found for this media type,
+    // prompt the user to enable one.
+    if (empty($display_options)) {
+      $form['attributes']['data-entity-view-mode-warning'] = [
+        '#markup' => '<div>' . t('There is no embedable view mode for this media type.') . '</div>',
```
Comment (sergepavle) [inline modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: $this->t

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 4]
```diff
@@ -336,12 +348,39 @@ public function buildEmbedStep(array $form, FormStateInterface $form_state): arr
...
+      ];
+      if ($this->moduleHandler->moduleExists('field_ui')) {
+        $form['attributes']['data-entity-view-mode-link'] = [
+          '#type' => 'link',
+          '#title' => t('Manage @media view modes', ['@media' => $entity->bundle()]),
```
Comment (sergepavle) [inline modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: $this->t

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 5]
```diff
@@ -672,19 +744,16 @@ protected function getMediaViewModeOptions(MediaInterface $media): array {
...
-      return [];
-    }
-
     $options = [];
     /** @var \Drupal\Core\Entity\Display\EntityViewDisplayInterface[] $displays */
     $displays = $this->entityTypeManager->getStorage('entity_view_display')->loadMultiple($displays);
     foreach ($displays as $display) {
-      $options[$display->getMode()] = $display->getMode() === 'default' ? $this->t('Default')->__toString() : $this->entityTypeManager->getStorage('entity_view_mode')->load('media.' . $display->getMode())->label();
+      if ($display->getThirdPartySetting('oe_media_embed', 'embeddable')) {
+        $options[$display->getMode()] = $display->getMode() === 'default' ? $this->t('Default')->__toString() : $this->entityTypeManager->getStorage('entity_view_mode')->load('media.' . $display->getMode())->label();
```
Comment (brummbar) [inline modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: is the `__toString()` needed? If yes, I prefer `(string)` as we do already around.

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 6]
```diff
@@ -658,6 +662,74 @@ protected function prepareAttributes(array $attributes, MediaInterface $media):
...
+  /**
+   * Returns the form element required to select a media view mode.
+   *
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The form state.
+   *
+   * @return array
+   *   The form element.
+   */
+  protected function getMediaViewModeFormElement(FormStateInterface $form_state) {
```
Comment (brummbar) [inline modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: We should add the needed parameters to this method, so `$entity_element` and `$entity`, as the form state is not really needed.

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php [hunk 7]
```diff
@@ -658,6 +662,74 @@ protected function prepareAttributes(array $attributes, MediaInterface $media):
...
+          '#attributes' => ['target' => '_blank'],
+        ];;
+      }
+
+      return $form_element;
+    }
+
+    // If there is only one display, use it and don't ask for input.
+    if (count($display_options) === 1) {
+      reset($display_options);
```
[the remark this answers is not in the document]
Comment (imanoleguskiza) [reply modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: But $display_options is an array and what we need is the value of the first (only) item

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php (hunk 6 above)
Comment (imanoleguskiza) [reply modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: We also need the embed_button from the form_state. Does that change anything? :D

File: modules/oe_media_embed/README.md [hunk 8]
```diff
@@ -2,7 +2,10 @@
 
 The OpenEuropa Media Embed module allows the embedding of Media entities into content in an agnostic (non-Drupal) way.
 
-To this end, it comes with two main elements: the WYSIWYG embed button and the filter plugin.
+To this end, it comes with two main elements: the WYSIWYG embed button and the filter plugin. On top of that, it forces
```
Comment (brummbar) [inline modules/oe_media_embed/README.md]: It allows site administrators, while it's true that if you don't pick any you cannot embed, "forces" sounds bad.

File: modules/oe_media_embed/README.md [hunk 9]
```diff
@@ -15,7 +18,39 @@ inserted in the content. This code follows the oEmbed protocol and is therefore
...
+
+## Text filter
+
+The embed code provided by the WYSIWYG button is transformed into the rendered entity by the filter plugin `FilterMediaEmbed`.
+Adding this to a text format will replace the embed tags with the rendered media entity.
+
+
+## Embeddable view modes
+
+Site adminstrators will need to define which available view modes are also available to be embedded via the tools described above.
```
Comment (brummbar) [inline modules/oe_media_embed/README.md]: "administrators"

File: modules/oe_media_embed/README.md [hunk 10]
```diff
@@ -15,7 +18,39 @@ inserted in the content. This code follows the oEmbed protocol and is therefore
...
+This is done by selecting the available view modes on the display mode configuration page for each available entity bundle.
```
Comment (brummbar) [inline modules/oe_media_embed/README.md]: media entity bundle

File: modules/oe_media_embed/README.md [hunk 11]
```diff
@@ -15,7 +18,39 @@ inserted in the content. This code follows the oEmbed protocol and is therefore
...
+
+## Usage
+
+In order to use the functionalities of the module, first you will need to create a text format.
+You can find more information on how to do that on the official [documentation][2]
+
+Once that is done you have to make sure that:
```
Comment (brummbar) [inline modules/oe_media_embed/README.md]: I'd use a step (ordered/unordered list) by step procedure here.

1. Create a text format can stay there where you put it;
2. Enable the CKEditor on it;
3. Add the media button to the toolbar;
4. Enable the text filter and place it last (very important as you wrote - perfect)
5. Go to the media display page (not content type) (provide example link) and enable which view modes. The embed view mode is shipped and enabled by default (is it?)

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php (hunk 7 above)
[the remark this answers is not in the document]
Comment (brummbar) [reply modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: Yep but `$display_options` has one element in there, right? But it's ok.

File: modules/oe_media_embed/src/Form/MediaEmbedDialog.php (hunk 6 above)
Comment (brummbar) [reply modules/oe_media_embed/src/Form/MediaEmbedDialog.php]: Where? I cannot see it :(

**Pull request merged**